src/bindle_utils.rs

```diff
@@ -1,24 +1,124 @@
 use itertools::Itertools;
+use std::collections::HashMap;
+use std::path::Path;
+
+use bindle::client::{
+    tokens::{HttpBasic, NoToken, TokenManager},
+    Client, ClientBuilder,
+};
+
+use crate::hippofacts::{HippoFacts, HippoFactsEntry};
+
+enum AuthMethod {
+    HttpBasic(HttpBasic),
+    None(NoToken),
+}
 
 pub struct BindleConnectionInfo {
     base_url: String,
     allow_insecure: bool,
+    auth_method: AuthMethod,
 }
 
 impl BindleConnectionInfo {
-    pub fn new<I: Into<String>>(base_url: I, allow_insecure: bool) -> Self {
+    pub fn new<I: Into<String>>(
+        base_url: I,
+        allow_insecure: bool,
+        username: Option<String>,
+        password: Option<String>,
+    ) -> Self {
+        let auth_method = match (username, password) {
+            (Some(u), Some(p)) => AuthMethod::HttpBasic(HttpBasic::new(&u, &p)),
+            _ => AuthMethod::None(NoToken::default()),
+        };
+
         Self {
             base_url: base_url.into(),
             allow_insecure,
+            auth_method,
         }
     }
 
-    pub fn client(&self) -> bindle::client::Result<bindle::client::Client> {
-        let options = bindle::client::ClientOptions {
-            http2_prior_knowledge: false,
-            danger_accept_invalid_certs: self.allow_insecure,
-        };
-        bindle::client::Client::new_with_options(&self.base_url, options)
+    pub fn client<T: TokenManager>(&self, token_manager: T) -> bindle::client::Result<Client<T>> {
+        let builder = ClientBuilder::default()
+            .http2_prior_knowledge(false)
+            .danger_accept_invalid_certs(self.allow_insecure);
+        builder.build(&self.base_url, token_manager)
+    }
+
+    pub async fn push_all(
```

Comment:
This method basically shows the issue I was dealing with: There is no way to pass around a `Client<dyn TokenManager>` or a `Client<Box<dyn TokenManager>>`. So instead, the `BindleConnectionInfo` now has a reference to the right `TokenManager`. However, instead of returning the client, it has to encapsulate the client for its lifespan. So I had to move some functionality into the `BindleConnectionInfo`.

There is probably some better way of doing it, but I couldn't figure out a way that didn't involve modifying Bindle.